File: Java/src/com/practice/Hard/ZumaGame.java
package com.practice.Hard;

/**
 * Think about Zuma Game. You have a row of balls on the table, colored red(R), yellow(Y), blue(B), green(G), and white(W). You also have several balls in your hand.
 *
 * Each time, you may choose a ball in your hand, and insert it into the row (including the leftmost place and rightmost place). Then, if there is a group of 3 or more balls in the same color touching, remove these balls. Keep doing this until no more balls can be removed.
 *
 * Find the minimal balls you have to insert to remove all the balls on the table. If you cannot remove all the balls, output -1.
 *
 * Examples:
 *
 * Input: "WRRBBW", "RB"
 * Output: -1
 * Explanation: WRRBBW -> WRR[R]BBW -> WBBW -> WBB[B]W -> WW
 *
 * Input: "WWRRBBWW", "WRBRW"
 * Output: 2
 * Explanation: WWRRBBWW -> WWRR[R]BBWW -> WWBBWW -> WWBB[B]WW -> WWWW -> empty
 *
 * Input:"G", "GGGGG"
 * Output: 2
 * Explanation: G -> G[G] -> GG[G] -> empty
 *
 * Input: "RBYYBBRRB", "YRBGB"
 * Output: 3
 * Explanation: RBYYBBRRB -> RBYY[Y]BBRRB -> RBBBRRB -> RRRB -> B -> B[B] -> BB[B] -> empty
 *
 * Note:
 * You may assume that the initial row of balls on the table won’t have any 3 or more consecutive balls with the same color.
 * The number of balls on the table won't exceed 20, and the string represents these balls is called "board" in the input.
 * The number of balls in your hand won't exceed 5, and the string represents these balls is called "hand" in the input.
 * Both input strings will be non-empty and only contain characters 'R','Y','B','G','W'.
 *
 *
 * you are allowed to do insert
 * remove those balls back if their number is 3+ and if they are of the same color
 *
 *
 */
public class ZumaGame {

    static final int MAXCOUNT = 6;   // the max balls you need will not exceed 6 since "The number of balls in your hand won't exceed 5"

    /**
     * Find the minimal balls you have to insert to remove all the balls on the table. If you cannot remove all the balls, output -1.
     *
     * the solution came from magtron_3, it has a great room for improvement
     *
     * this problem has some issues by design.
     * @param board
     * @param hand
     * @return
     */
    public int findMinStep(String board, String hand) {
        int[] colorCount = new int[26];
        for (int i = 0; i < hand.length(); ++i)
            ++colorCount[hand.charAt(i) - 'A'];
        int rs = helper(board + "#", colorCount);  // append a "#" to avoid special process while j==board.length, make the code shorter.
        return rs == MAXCOUNT ? -1 : rs;
    }
    private int helper(String s, int[] h) {
        s = removeConsecutiveColors(s);
        if (s.equals("#")) return 0;

        int res = MAXCOUNT, qualified = 0;

        // i holds, j runs
        for (int i = 0, j = 0 ; j < s.length(); ++j) {
            if (s.charAt(j) == s.charAt(i)) continue;
            qualified = 3 - (j - i);     //balls need to be removed from current consecutive balls.
            if (h[s.charAt(i) - 'A'] >= qualified) {

                // update the global counter
                h[s.charAt(i) - 'A'] -= qualified;

                // problem reduction happens here
                res = Math.min(res, qualified + helper(s.substring(0, i) + s.substring(j), h));

                // tailspin compensation
                h[s.charAt(i) - 'A'] += qualified;
            }

            // catch up, maintaining the two pointers
            i = j;
        }
        return res;
    }

    private String removeConsecutiveColors(String board) {
        for (int i = 0, j = 0; j <= board.length(); ++j) {

            if (j < board.length() && board.charAt(j) == board.charAt(i))
                continue;

            if (j - i >= 3)
                return removeConsecutiveColors(board.substring(0, i) + board.substring(j));
            // yet another catch up
            i = j;
        }
        return board;
    }
}

File: Python/python3/leetcode/Hard/ZumaGame.py
# 488
import sys


# You are playing a variation of the game Zuma.
#
# In this variation of Zuma, there is a single row of colored balls on a board, where each ball can be colored red 'R', yellow 'Y', blue 'B', green 'G', or white 'W'. You also have several colored balls in your hand.
#
# Your goal is to clear all of the balls from the board. On each turn:
#
# Pick any ball from your hand and insert it in between two balls in the row or on either end of the row.
# If there is a group of three or more consecutive balls of the same color, remove the group of balls from the board.
# If this removal causes more groups of three or more of the same color to form, then continue removing each group until there are none left.
# If there are no more balls on the board, then you win the game.
# Repeat this process until you either win or do not have any more balls in your hand.
# Given a string board, representing the row of balls on the board, and a string hand, representing the balls in your hand, return the minimum number of balls you have to insert to clear all the balls from the board. If you cannot clear all the balls from the board using the balls in your hand, return -1.

class ZumaGame:

    def findMinStep(self, board, hand):
        def reduce_three_plus(brd): # 3 or 3+ consecutive chars will be reduced to a single char
            st = []
            n = len(brd)
            for i in range(n):
                ch = brd[i]
                st.append(ch)
                if (i == n - 1 or ch != brd[i + 1]) and len(st) >= 3:
                    a, b, c = st.pop(), st.pop(), st.pop()
                    if a == b == c:
                        while st and st[-1] == a:
                            st.pop()
                    else:
                        st.append(c)
                        st.append(b)
                        st.append(a)
            return ''.join(reversed(st))

        def dfs(brd, hd):
            key = brd + "_" + hd
            if key in d:
                return d[key]

            brd = reduce_three_plus(brd)
            ret, bsize, hsize = sys.maxsize, len(brd), len(hd)
            if bsize == 0 or hsize == 0:
                d[key] = 0 if bsize == 0 else sys.maxsize
                return 0 if bsize == 0 else sys.maxsize

            for i in range(hsize):
                for j in range(bsize):
                    if brd[j] == hd[i] or (j <= bsize - 2 and brd[j] == brd[j + 1]):
                        # run dfs on new board and new hand
                        ret = min(ret, 1 + dfs(brd[:j + 1] + hd[i] + brd[j + 1:], hd[:i] + hd[i + 1:]))

            d[key] = ret
            return ret
        
        d = {}
        ans = dfs(board, hand)
        return -1 if ans >= sys.maxsize else ans


board = "WRRBBW"
hand = "RB"
actual = ZumaGame().findMinStep(board, hand)
expected = -1
assert(expected == actual)

board = "WWRRBBWW"
hand = "WRBRW"
actual = ZumaGame().findMinStep(board, hand)
expected = 2
assert(expected == actual)


board = "G"
hand = "GGGGG"
actual = ZumaGame().findMinStep(board, hand)
expected = 2
assert(expected == actual)
